### ISSUE TITLE
4.1.0 don't connect to device

### ISSUE THREAD

**Issue by jmcollin78** [rospogrigio/localtuya#1077]
## The problem
Since 4.1.0 I've got many connection problems with my devices which were correctly working with 4.0.2.

## Environment

- Localtuya version: 4.1.0
- Last working localtuya version (if known and relevant): 4.0.2
- Home Assistant Core version: 2022.10.1 
- [x] Are you using the Home Assistant Tuya Cloud component ? 
- [x] Are you using the Tuya App in parallel ? 

## Steps to reproduce

1. Many devices are disconnected,
2. Check the logs and see many error messages

## Configuration `configuration.yaml` or `config_flow`

Not relevant

## DP dump

Not relevant

## Provide Home Assistant taceback/logs

```txt
2022-10-11 08:18:46.715 ERROR (MainThread) [custom_components.localtuya.common] [260...56c] Connect to 192.168.1.87 failed
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 390, in acquire
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/localtuya/common.py", line 216, in _make_connection
    await self._interface.reset(self._default_reset_dpids)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 521, in reset
    return await self.exchange(RESET, dpIds)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 486, in exchange
    msg = await self.dispatcher.wait_for(seqno)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 259, in wait_for
    await asyncio.wait_for(self.listeners[seqno].acquire(), timeout=timeout)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
```

```txt
2022-10-11 08:28:03.654 ERROR (MainThread) [custom_components.localtuya.common] [260...56c] Failed to set DP 1 to False
Traceback (most recent call last):
  File "/config/custom_components/localtuya/common.py", line 316, in set_dp
    await self._interface.set_dp(state, dp_index)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 553, in set_dp
    return await self.exchange(SET, {str(dp_index): value})
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 492, in exchange
    payload = self._decode_payload(msg.payload)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 623, in _decode_payload
    return json.loads(payload)
  File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
## Additional information
To recover my devices, a restart of Home assistant is NOT sufficient. I have to reboot electrically all my devices !!

**Comment by rospogrigio** [comment]
Just released a new version (4.1.1), try to update and see if it fixes your issues...

**Comment by jmcollin78** [comment]
I'm not so confident now... I have explosed the WAF (no more heating nor light yesterday)

**Comment by remark1973** [comment]
Home Assistant 2022.10.3
Supervisor 2022.10.0
Operating System 9.2
Frontend 20221010.0 - latest

Localtuya version: 4.1.1 

None of the 6 Wi-Fi bulbs I have are available in HA

**Comment by hitnrun30** [comment]
I noticed this too. Certain devices that have been fine forever are all of a sudden disconnecting.

**Comment by AFumi39** [comment]
Me too, I had to rollback to 4.0.2, because 4.1.1 don't work at all. It works for few minutes after reboot, then it suddenly fails to connect to all the devices (and then all the devices goes offline).

**Comment by AFumi39** [comment]
Correction: neither the 4.0.2 now works... After some minutes all the devices go offline. How's this possible?

**Comment by rospogrigio** [comment]
Are you sure you are not having network issues? Does the SmartLife app work? Do you see all the devices online in there?

**Comment by AFumi39** [comment]
> Are you sure you are not having network issues? Does the SmartLife app work? Do you see all the devices online in there?

Su, no network issues, all the other devices working fine. Tuya app works fine with the devices.
I've already restarted the Access Point.
I'm trying the last thing with the 4.1.1 to see if works. I'll let you know if I find something useful

**Comment by hitnrun30** [comment]
> Ok, now it works with 4.1.1 and OS 9.2 But you can't add or edit devices, or the integration stop working (I don't know why)

Yeah, all of a sudden it is working better.  Maybe it was the Tuya side that was messing up

**Comment by AFumi39** [comment]
With 4.1.1 I've still some problems. Sometimes it goes offline and took some minutes to go online again. That's not very big problem, but this morning (from about 9:00) it continually goes offline-online (every few minutes).
Another issue is that I have 1 smart plug that is online in Tuya, but offline from days in LocalTuya. Another one gone offline at 09:04 (and still offline now, at 09:51), but is ok in Tuya app.
That's very strange

**Comment by rospogrigio** [comment]
Can you all try PR #1087 ? @sibowler has provided it and we'd need some feedback to see if it fixes these issues that arose since 4.1.0. Thank you

**Comment by AFumi39** [comment]
> Can you all try PR #1087 ? @sibowler has provided it and we'd need some feedback to see if it fixes these issues that arose since 4.1.0. Thank you

After a reboot it seems to work again flawlessly (so it's probably not a device/tuya problem, but something inside LocalTuya that breaks after some event).
I've changed the lines in that commit, I'll let you know if it solves the problem or not (if it works, I've to wait some days to be sure).

Thanks

**Comment by rospogrigio** [comment]
Well something that works intermittently could very likely be related to ha semaphore not handled correctly so maybe @sibowler is on the right track. OK let us know.

**Comment by AFumi39** [comment]
> Well something that works intermittently could very likely be related to ha semaphore not handled correctly so maybe @sibowler is on the right track. OK let us know.

I've just posted in the PR thread an error related to DPS of one device that now results offline in HA (cause of this error, I suppose). I think that this could be the main issue that occurs, but not correctly catched before the semaphore fix... maybe.

Thank you

**Comment by syndikatss** [comment]
Hello, I am writing through a translator on this text, it may seem a little silly) On versions 4.1.0 and 4.1.1, there are problems with connecting the relay, it is de-energized for me and works only when the light in the room turns on, out of 10 connections it can work once or not earn at all. At the same time, everything works fine in the application. If required, I can install a new version and see if there are any errors in the home assistant.

**Comment by jmcollin78** [comment]
I rollback to 4.0.2 and some devices are still unavailable through LocalTuya but works fine with official Tuya and with the app.
I try to remove and re-add the device but when trying to re-add the device I've got this error:
![Capture d’écran 2022-10-19 à 06 53 54](https://user-images.githubusercontent.com/1712374/196600761-2bb6678c-cf20-4f24-8d63-f4372baf810e.png)

The error log always says the same error:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 390, in acquire
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/localtuya/common.py", line 170, in _make_connection
    status = await self._interface.status()
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 481, in status
    status = await self.exchange(STATUS)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 460, in exchange
    msg = await self.dispatcher.wait_for(seqno)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 247, in wait_for
    await asyncio.wait_for(self.listeners[seqno].acquire(), timeout=timeout)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
```

I try reboot the device, reboot HA, remove and add the device in integration with no luck.
I'm totally stuck with this 2 devices.
Something that may be related, those 2 devices have been upgraded a few days before (OTA firmware update).

Is there something I can look in the code or change for trying to resolve this problem ?

**Comment by jmcollin78** [comment]
I will try 4.1.1 and let you know

**Comment by jmcollin78** [comment]
With 4.1.1 I still can't connect to the device but the error is now different:
![Capture d’écran 2022-10-19 à 07 16 06](https://user-images.githubusercontent.com/1712374/196603334-b0181179-874d-4d95-aa76-cf03e044819b.png)

In the log I can see that:
```
2022-10-19 07:21:55.656 ERROR (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Failed to get status: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 390, in acquire
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 574, in detect_available_dps
    data = await self.status()
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 507, in status
    status = await self.exchange(STATUS)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 486, in exchange
    msg = await self.dispatcher.wait_for(seqno)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 259, in wait_for
    await asyncio.wait_for(self.listeners[seqno].acquire(), timeout=timeout)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
```

**Comment by jmcollin78** [comment]
I try the fix described here : https://github.com/rospogrigio/localtuya/pull/1087 with no more success.

**Comment by remark1973** [comment]
> Home Assistant 2022.10.3 Supervisor 2022.10.0 Operating System 9.2 Frontend 20221010.0 - latest
> 
> Localtuya version: 4.1.1
> 
> None of the 6 Wi-Fi bulbs I have are available in HA

In my case Tuya IOT API was expired. After API trial period prolongation, deleting and adding devices solved my problem

**Comment by jmcollin78** [comment]
In 4.1.1 after a few hours, many of my lights turns to Disabled and the log shows this:

```
2022-10-19 12:35:59.434 ERROR (MainThread) [custom_components.localtuya.pytuya] [322...29d] Failed to get status: 'NoneType' object has no attribute 'write'
Traceback (most recent call last):
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 581, in detect_available_dps
    data = await self.status()
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 514, in status
    status = await self.exchange(STATUS)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 492, in exchange
    self.transport.write(payload)
AttributeError: 'NoneType' object has no attribute 'write'
```

Any help would be greatly appreciated. I'm totally stuck with 4.1.1, partially stuck with 4.0.2 now.
I don't what to do to recover a normal working way

**Comment by AFumi39** [comment]
> Well something that works intermittently could very likely be related to ha semaphore not handled correctly so maybe @sibowler is on the right track. OK let us know.

Unfortunately it don't resolve the issue completely. Some device gone offline tonight at 06:00 AM, more or less, (all together) and after lot of try probably HA stopped the requests from LucalTuya after 6 hours (at about 12:00). So the devices figures all offline.

After reboot of HA, all works again, except a single device (different from the last one) that now have the "unknown DPS" error.
That's very strange

**Comment by sibowler** [comment]
@jmcollin78 - Those recent Timeout errors (including the one on 4.0.2) seem to indicate a connection failure. Could the key have changed? IF those devices have been updated OTA possibly something has changed in the new firmware that breaks the localtuya approach. 

Re your "AttributeError: 'NoneType' object has no attribute 'write'" errors for the lights - have you configured any of the RESET or polling device parameters at all? Also do you have these devices setup on both the native Tuya integration AND the localtuya one? I've found that often the devices can only connect to one 'controller', i.e. opening the TuyaSmart app will cause disconnections in localtuya for those devices. Any chance that might be occurring in your network? It also sounds like you've not got these blocked from the internet - so not sure whether the device connecting to TuyaCloud may also cause issues here.

**Comment by jmcollin78** [comment]
Hello,
@sibowler Thank you for your suggestion. 

The localkey doesn't have change, I have not configured any RESET or polling.

I have the offical Tuya and localtuya configured both. Until last week it have never been a problem, then I will try this clue.

fyi, yesterday after a HA reboot, lights become up again. So something breaks but I have no idea of what happens.
This night the cover closes itself alone, nobody and no automation have runs... There are gremlins in my house since few days which coincide with an electrical outage and the installation of 4.1.0. I don't remember in what order those 2 events  happens, but the electrical outage have lead to many connection problems. I have to reboot HA and certain devices to make it works again.

I have seen on a forum that it is not necessary to set the DHCP static reservation. All my devices have a static DHCP IP address. If my DHCP reservations have expired during the outage, is it possible that Localtuya is using a "wrong" ip address (and so the IP connection is done but the device is not responding as expected or something like that).

**Comment by jmcollin78** [comment]
> IF those devices have been updated OTA possibly something has changed in the new firmware that breaks the localtuya approach.

That is my opinion. What is this approach that could be broken ? This is Bseed socket with power monitoring like this: https://www.amazon.fr/BSEED-Intelligente-alternatif-Compatible-standard/dp/B08MVZFST6/ref=sr_1_6?__mk_fr_FR=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=3UJPH5VPITLGB&keywords=bseed+prise+connect%C3%A9e+wifi&qid=1666252484&sprefix=bseed+prise+connect%C3%A9e+wifi%2Caps%2C50&sr=8-6 and I guess they are relatively popular so may be others will have the same trouble. I'm ok to debug and help trying to find what happens.
May be I can open another case for this specific issue (which is not related to 4.1.1). If someone can help me (i'm a developper but not an expert of HA, nor python), i'm ready to start debugging this case.

**Comment by jmcollin78** [comment]
I suppress my 2 sockets from Smart Home, suppress from HA, reboot HA, re-add in Smart App, try to add in LocalTuya and then I've got this error:
```
2022-10-20 11:02:47.852 ERROR (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Failed to get status: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 390, in acquire
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 581, in detect_available_dps
    data = await self.status()
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 514, in status
    status = await self.exchange(STATUS)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 493, in exchange
    msg = await self.dispatcher.wait_for(seqno)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 259, in wait_for
    await asyncio.wait_for(self.listeners[seqno].acquire(), timeout=timeout)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
```

Do you agree that it is not possible to add a device in LocalTuya if this device is NOT already in the official Tuya integration ? Because we need the local_key, the device should be paired in Smart Home and then should be in official Tuya application. Am I wrong ?

**Comment by rospogrigio** [comment]
Yes, of course you need to add the device to your SmartHome app first, or you won't be able to get the local_key (which changes every time you pair the device to the app).

**Comment by jmcollin78** [comment]
I put localtuya in debug mode and here is what I see:

```
2022-10-20 11:37:30.573 DEBUG (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Sending command status (device type: type_0a)
2022-10-20 11:37:30.573 DEBUG (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Send payload: b'{"gwId":"bfe407813ac1646f5236tb","devId":"bfe407813ac1646f5236tb"}'
2022-10-20 11:37:30.576 DEBUG (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Waiting for sequence number 0
2022-10-20 11:37:30.691 DEBUG (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Connection lost: None
2022-10-20 11:37:35.578 ERROR (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Failed to get status: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 390, in acquire
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 581, in detect_available_dps
    data = await self.status()
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 514, in status
    status = await self.exchange(STATUS)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 493, in exchange
    msg = await self.dispatcher.wait_for(seqno)
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 259, in wait_for
    await asyncio.wait_for(self.listeners[seqno].acquire(), timeout=timeout)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
2022-10-20 11:37:35.585 DEBUG (MainThread) [custom_components.localtuya.config_flow] Initial state update failed, trying reset command
2022-10-20 11:37:35.585 DEBUG (MainThread) [custom_components.localtuya.config_flow] No DPS able to be detected
2022-10-20 11:37:35.585 DEBUG (MainThread) [custom_components.localtuya.config_flow] Detected DPS: {}
2022-10-20 11:37:35.585 DEBUG (MainThread) [custom_components.localtuya.pytuya] [bfe...6tb] Closing connection
```

**Comment by jmcollin78** [comment]
No more problems since a few weeks. I close

**Comment by hitnrun30** [comment]
Actually of lately the devices keep going unavailable and turning on when they become available
